server/data/data-jpa/src/main/java/org/apache/james/sieve/jpa/JPASieveRepository.java

```diff
@@ -0,0 +1,391 @@
+                .filter(sieveScript -> !sieveScript.getScriptName().equals(name.getValue()))
+                .mapToLong(JPASieveScript::getScriptSize)
+                .sum();
+
+        QuotaSize quota = limitToUse(user);
+        if (overQuotaAfterModification(usedSpace, size, quota)) {
+            throw new QuotaExceededException();
+        }
+    }
+
+    private QuotaSize limitToUse(User user) throws StorageException {
+        Optional<JPASieveQuota> userQuota = findQuotaForUser(user.asString());
+        if (userQuota.isPresent()) {
+            return QuotaSize.size(userQuota.map(JPASieveQuota::getSize));
+        }
+
+        Optional<JPASieveQuota> defaultQuota = findQuotaForUser(DEFAULT_SIEVE_QUOTA_USERNAME);
+        if (defaultQuota.isPresent()) {
+            return QuotaSize.size(defaultQuota.map(JPASieveQuota::getSize));
+        } else {
+            return QuotaSize.unlimited();
+        }
+    }
+
+    private boolean overQuotaAfterModification(final long usedSpace, final long size, final QuotaSize quota) {
+        return QuotaSize.size(usedSpace).add(size).isGreaterThan(quota);
+    }
+
+    @Override
+    public void putScript(final User user, final ScriptName name, final ScriptContent content) throws StorageException, QuotaExceededException {
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        EntityTransaction transaction = entityManager.getTransaction();
+        try {
+            transaction.begin();
+
+            haveSpace(user, name, content.length());
+            JPASieveScript jpaSieveScript = JPASieveScript.builder(user.asString(), name.getValue()).scriptContent(content).build();
+            entityManager.persist(jpaSieveScript);
+
+            transaction.commit();
+        } catch (SieveRepositoryException e) {
+            rollbackTransactionIfActive(transaction);
+            throw e;
+        } catch (PersistenceException e) {
+            rollbackTransactionIfActive(transaction);
+            throw new StorageException("Unable to put script for user " + user.asString(), e);
+        } finally {
+            entityManager.close();
+        }
+    }
+
+    @Override
+    public List<ScriptSummary> listScripts(final User user) throws StorageException {
+        return findAllSieveScriptsForUser(user).stream()
+                .map(sieveScript -> new ScriptSummary(new ScriptName(sieveScript.getScriptName()), sieveScript.isActive()))
+                .collect(Collectors.toList());
+    }
+
+    private List<JPASieveScript> findAllSieveScriptsForUser(final User user) throws StorageException {
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        try {
+            List<JPASieveScript> sieveScripts = entityManager.createNamedQuery("findAllByUsername", JPASieveScript.class)
+                    .setParameter("username", user.asString()).getResultList();
+            return sieveScripts != null ? sieveScripts : new ArrayList<>();
```

Comment:
We tend to use Optionals for null checks.

BTW prefer `ImmutableList.of()` over `new ArrayList<>()`